feat: Parallel prefill (#846)

jthomson04 opened the pull request:
#### Overview:

Enable parallel prefill for disagg serving. The updated process is to dequeue multiple requests until a token threshold is reached. Then, all requests are completed before dequeueing a new set of requests.

- closes GitHub issue: https://github.com/ai-dynamo/dynamo/issues/846


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced batch processing for prefill requests, allowing multiple requests to be handled together for improved efficiency.
  - Added a configurable parameter to set the maximum number of tokens processed in a prefill batch via command-line arguments and configuration files.

- **Improvements**
  - Enhanced queue handling to support batch dequeuing of prefill requests based on token limits.
  - Added timeout support for dequeuing individual prefill requests.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jthomson04 committed:
Parallel prefill

PeaBrane commented:
Would it make more sense for the batching logic to be refactored into `utils/prefill_queue.py` instead? As currently `prefill_worker.py` looks quite nested. 
Maybe something like `dequeue_prefill_request` would take the optional kwarg `max_batched_tokens` etc

Is it also worth it to have an additional kwarg specifying an optional time duration such that the batch would be spit out regardless of whether it has hit the token limit, in the case where we don't get reqs for a long period of time and the entire thing just hangs (?)

PeaBrane commented on examples/llm/components/prefill_worker.py:
```diff
@@ -133,13 +141,40 @@ async def prefill_queue_handler(self):
-                        pass
+
+                    # TODO: We might want to double-buffer this process
+                    # to avoid the overhead of dequeuing from nats
```
the logic here is a bit messy IMO, I think a guard clause checking the token limit and breaking immediately would avoid some nestings here 
also, currently I think it's possible for the "last" req to just get dropped?

PeaBrane replied:
I think we may need to eventually run this through the pipeline to verify correctness (the correct # of requests get processed etc), and performance is as expected

tedzhouhk replied:
Agreed, may I propose to move this logic to python PrefillQueue? i.e., have a function like `PrefillQueue.dequeue_n_reqs_until_(absolute_)isl_is_larger_than(isl)`. This can simply the execution loop to 

```
while True:
    reqs = PrefillQueue.dequeue_n_reqs_until_(absolute_)isl_is_larger_than(isl)
    futures = [wrap_generator(self.generate(req)) for req in reqs]
    await asyncio.gather(*futures)
```

jthomson04 commented:
> Would it make more sense for the batching logic to be refactored into `utils/prefill_queue.py` instead? As currently `prefill_worker.py` looks quite nested. Maybe something like `dequeue_prefill_request` would take the optional kwarg `max_batched_tokens` etc

Yes, I'll refactor.

> Is it also worth it to have an additional kwarg specifying an optional time duration such that the batch would be spit out regardless of whether it has hit the token limit, in the case where we don't get reqs for a long period of time and the entire thing just hangs (?)

I don't think a timeout would be needed here. That's what the check for the size of the queue is for. After waiting an arbitrary time for the first request to come in and form the batch, we ensure that the max amount of requests we try to dequeue is the same as the remaining queue size.

PeaBrane commented:
> I don't think a timeout would be needed here. That's what the check for the size of the queue is for. After waiting an arbitrary time for the first request to come in and form the batch, we ensure that the max amount of requests we try to dequeue is the same as the remaining queue size.

Ah yes, makes sense. And if the queue is empty we'd just break.

tedzhouhk commented on examples/llm/components/prefill_worker.py:
```diff
@@ -133,13 +141,40 @@ async def prefill_queue_handler(self):
+                    # If a single request is larger than the threshold, we will execute it as bs=1.
+                    remaining_prefill_tokens = (
+                        self.engine_args.max_batched_prefill_tokens
+                        - len(req.prompt_token_ids)
```
Should we use absolute isl? (i.e., `len(req.prompt_token_ids) - len(req.computed_block_ids) * engine_args.block_size`) `req.computed_block_ids` will be read from decode and is only attended to (no dense operation). In my previous benchmark, prefill time is dominant by absolute isl and prefix cache doesn't affect prefill time much.

tedzhouhk commented on examples/llm/utils/vllm.py:
```diff
@@ -69,6 +69,12 @@ def parse_vllm_args(service_name, prefix) -> AsyncEngineArgs:
         default=3,
         help="Maximum queue size for remote prefill. If the prefill queue size is greater than this value, prefill phase of the incoming request will be executed locally.",
     )
+    parser.add_argument(
+        "--max-batched-prefill-tokens",
+        type=int,
+        default=2048,
+        help="Maximum number of tokens to prefill in a single batch. If the number of tokens to prefill is greater than this value, prefill phase will execute as bs=1.",
+    )
```
> Maximum number of tokens to prefill in a single batch

Is this accurate name/description? With nats queue, we can only ensure stop dequeuing when batched prefill tokens is larger than this value. Or, the actual num prefill token is always larger than this value.

jthomson04 replied:
Alternatively, this could be the max, where we dequeue an extra one (that we find exceeds the max) and carry it over to the next batch.

tedzhouhk replied:
I don't think that's a good idea. We should send whatever we dequeue to the engine immediately, or we shouldn't dequeue and hope other prefill workers will pick it up. 

How about naming it `--min-batched-prefill-tokens` and explain that this only applies if the prefill queue has that many tokens.

tedzhouhk commented on examples/llm/components/prefill_worker.py:
```diff
@@ -133,13 +141,40 @@ async def prefill_queue_handler(self):
+                        prefill_queue_size = await prefill_queue.get_queue_size()
+                        for _ in range(prefill_queue_size):
+                            # This should be dequeued almost immediately, hence the very short timeout.
+                            req = await prefill_queue.dequeue_prefill_request(0.1)
```
Is 0.1s too long? This is adding 100ms to TTFT in low-latency cases. We would want something to return almost immediately

jthomson04 replied:
Since we call dequeue knowing that something's already in the queue, this would return almost immediately

tedzhouhk replied:
Hmm, you're right. But it could still add 100ms in rare cases, as this is multi-processes, we can't guarantee the PrefillQueue won't change after checking it's size

jthomson04 replied:
Ah ok. Will reduce to 0.01

tedzhouhk replied:
How about `while(await prefill_queue.dequeue(0))`?

jthomson04 replied:
Seems like setting timeout 0 leads the batch size to always be 1, even when the size of the queue is much larger, so we'll need some non-zero value.

jthomson04 committed:
Improve parallelism + refactor + bugfixes

jthomson04 committed:
Use absolute isl

jthomson04 commented:
Refactored the batching logic into prefill queue, fixed bug that could cause some requests to be dropped, and transitioned to using absolute ISL. The current implementation pulls requests until the sum of the absolute isls exceeds the threshold. It then takes the final request that pushed it over the threshold, and puts it aside to be used as the first request of the next batch.

PeaBrane commented on examples/llm/utils/prefill_queue.py:
```diff
@@ -39,18 +39,67 @@ def __init__(
+        remaining_prefill_tokens = max_batched_prefill_tokens - num_new_tokens(req)
+
+        if remaining_prefill_tokens < 0:
+            return reqs
```
I think you need to `self.pending = req` here, otherwise you still drop it?

Since you are already returning reqs directly here as a guard, you don't need the else clause, doesn't change anything, just less nesting

PeaBrane replied:
I think you need to `self.pending = req` here, otherwise you still drop it?

Since you are already returning reqs directly here as a guard, you don't need the else clause, doesn't change anything, just less nesting

jthomson04 committed:
Little refactor

PeaBrane commented:
I think it LGTM. @tedzhouhk @jthomson04 do you guys have a machine to run a genai-perf with this, and I think we are 👍.  (Verify num requests, avg ISL / OSL as expected etc)

tedzhouhk commented:
> I think it LGTM. @tedzhouhk @jthomson04 do you guys have a machine to run a genai-perf with this, and I think we are 👍. (Verify num requests, avg ISL / OSL as expected etc)

No I don't have a machine now 😢 how about we default max-batched-prefill-tokens (or whatever it will be called) to 1 and set to 1 in the configs so that default behavior keeps the same?

tedzhouhk commented on examples/llm/components/prefill_worker.py:
```diff
@@ -39,6 +39,14 @@ class RequestType(BaseModel):
     text: str
 
 
+async def wrap_generator(generator):
+    while True:
+        try:
+            await generator.__anext__()
+        except StopAsyncIteration:
+            break
+
+
```
nit: how about define this inside `PrefillWorker` class so that if we have a batched generate in the future we can direct swap?

jthomson04 replied:
Hmm. That would mean we'd also need to bring the PrefillWorker.generate call into the prefill queue, which might get a bit weird.

jthomson04 commented:
> I think it LGTM. @tedzhouhk @jthomson04 do you guys have a machine to run a genai-perf with this, and I think we are 👍. (Verify num requests, avg ISL / OSL as expected etc)

I can do benchmarks on A6000 sometime tomorrow

jthomson04 committed:
Merge remote-tracking branch 'origin/main' into jthomson04/parallel-prefill2

jthomson04 committed:
Merge branch 'main' into jthomson04/parallel-prefill2

jthomson04 commented:
Some unfortunate findings on the benchmarks: The batched prefills with the current approach have a TTFT from 1-30% longer than the prior approach. So far, I've only tested on genai-perf at ISL 1000, OSL 256, and concurrencies 5, 10, and 20.

tedzhouhk commented:
> Some unfortunate findings on the benchmarks: The batched prefills with the current approach have a TTFT from 1-30% longer than the prior approach. So far, I've only tested on genai-perf at ISL 1000, OSL 256, and concurrencies 5, 10, and 20.

@jthomson04 We need prefill to be short and lagging (i.e., cannot catch up with input requests) to be able to see the improvement. I would suggest try with shorter isl and more requests, i.e., isl=250, num_req=100-500 dumped all at t=0 (you can achieve this by setting conc=num_requests in GAP). Also I recommend turning off conditional remote prefill for benchmarking to make sure prefill engine is doing all the prefill.

jthomson04 committed:
Merge branch 'main' into jthomson04/parallel-prefill2

jthomson04 commented:
Updated benchmarks show a significant reduction in TTFT, as well as a slight increase in output token throughput with the batched prefills.

jthomson04 committed:
Merge branch 'main' into jthomson04/parallel-prefill2

jthomson04 committed:
Merge branch 'main' into jthomson04/parallel-prefill2

jthomson04 committed:
Merge branch 'main' into jthomson04/parallel-prefill2

jthomson04 committed:
Merge branch 'main' into jthomson04/parallel-prefill2

rmccorm4 commented:
Closing due to lack of activity
